src/rust/engine/fs/fs_util/src/main.rs

```diff
@@ -224,11 +227,14 @@ fn execute(top_match: clap::ArgMatches) -> Result<(), ExitError> {
           match file {
             fs::Stat::File(f) => {
               let digest = FileSaver {
-                store: store,
+                store: store.clone(),
                 posix_fs: Arc::new(posix_fs),
               }.digest(&f)
                 .wait()
                 .unwrap();
+              if store_has_remote {
```

Comment:
It seems like it would be much easier to make `ensure_remote_has_recursive` a noop in the case where there is no remote store, and not pass this boolean around? But there was a reason why we didn't do that I've no forgotten I think.

Reply:
Yeah; there are two styles of caller here:

1. The process execution caller, which calls `ensure_remote_has_recursive` because it specifically knows that files are missing, and needs to know that action has actually been taken as a result of its call. In this case, a no-op `ensure_remote_has_recursive` (because a poorly configured `Store` has been passed) likely leads to an infinite loop of the executor asking for files to be uploaded and retrying the action. What we really want to do in this case is error hard, saying "Files were missing, they couldn't be uploaded".
2. The fs_util caller, who calls `ensure_remote_has_recursive` because it wants to proactively ensure this is the case.

I'd rather optimise for the first case, which is in meaningful complicated code, than the second case (where this boolean is actually very easy to create and pass around, and its scope is very limited).